### PR TITLE
Captured args are resolved in kernel definition scope

### PR DESCRIPTION
**Pull request by lmondada**
Python follows lexical scoping, which means that captured variables in a function `foo` must resolve in the scope where `foo` is defined, and not in the caller's scope. This is different than arguments passed to a function, which are (obviously) resolved in the caller's scope. For instance:

```python
# Module mod_a.py

a = 3
def foo():
    print(a)

# Module mod_b.py
from mod_a import foo

a = "something else"
foo()
```

should print `3`.

This PR adjusts the behaviour of captured arguments in kernels to behave similarly, by keeping a reference to the frame that the kernel was defined in.

See tests and https://github.com/NVIDIA/cuda-quantum/issues/3965.

closes https://github.com/NVIDIA/cuda-quantum/issues/3965

### CODE REVIEW

File: python/cudaq/kernel/kernel_decorator.py [hunk 1]
```diff
@@ -536,16 +519,37 @@ def handle_call_arguments(self, *args):
...
+        Resolve the arguments passed to the decorator at call site.
+        """
+        return [
+            self.process_argument(arg, arg_type)
+            for arg, arg_type in zip(args, self.arg_types())
+        ]
+
+    def prepare_call(self, *args):
+        """
+        Process call site arguments, capture lifted arguments and retrieve
```
Comment (bettinaheim) [inline python/cudaq/kernel/kernel_decorator.py:548]: nit: worth capturing in the doc comment that this creates a clone of the qkeModule.

File: python/tests/kernel/test_deferred_compilation.py [hunk 2]
```diff
@@ -221,6 +221,157 @@ def bar():
...
+
+
+def _make_caller_callee_pair():
+    """Helper: define a caller/callee pair inside a nested scope and return them."""
+
+    @cudaq.kernel
+    def inner(q: cudaq.qubit):
+        x(q)
+
+    @cudaq.kernel
```
Comment (bettinaheim) [inline python/tests/kernel/test_deferred_compilation.py:236]: nit: Edge case that is not properly dealt with:
using the `@kernel` attribute in an inner function currently I believe will be accepted by the bridge (the attribute is simply ignored). I believe the logic to get the correct frame relies on functions defined within kernels not having that attribute. That's fine - but if it is a quick thing to do it would be nice to make the bridge give an error if you do this:
```
    @cudaq.kernel
    def kernel(q: cudaq.qubit):
        @cudaq.kernel # error?
        def inner_fct():
            ...
```

File: python/cudaq/kernel/ast_bridge.py [hunk 3]
```diff
@@ -3909,9 +3886,11 @@ def get_item_type(pyval):
                 return cc.StdvecType.get(base_elTy)
             elif isinstance(pyval, ast.Call):
                 if isinstance(pyval.func, ast.Name):
+                    from .kernel_decorator import isa_kernel_decorator
                     # supported for calls but not here: 'range', 'enumerate'
-                    decorator = recover_kernel_decorator(pyval.func.id)
-                    if decorator:
+                    decorator = recover_value_of_or_none(
```
Comment (schweitzpgi) [inline python/cudaq/kernel/ast_bridge.py:3918]: Why are you getting rid of the encapsulation of this functionality to find the decorator?

**Comment by schweitzpgi** [comment]
What would be helpful for reviews is to avoid gratuitous symbol name changes in the PR and focus on functional changes or (exclusively or) perfecting the spelling of symbols in the code in separate PRs. That will let me focus on how you are changing things rather than spending bandwidth tripping over spelling changes. Thanx.

File: python/cudaq/kernel/kernel_builder.py [hunk 4]
```diff
@@ -1374,11 +1373,6 @@ def resolve_callable_arg(self, insPt, target):
         fn = recover_func_op(self.module, fulluniq)
 
         # build the closure to capture the lifted `args`
-        thisPyMod = recover_calling_module()
```
Comment (schweitzpgi) [inline python/cudaq/kernel/kernel_builder.py:1377]: Need to discuss this offline.

File: python/cudaq/kernel/ast_bridge.py (hunk 3 above)
Comment (lmondada) [reply python/cudaq/kernel/ast_bridge.py:3918]: `recover_kernel_decorator` does the same thing as `recover_value_of_or_none`. I have just consolidated and removed one of the two redundant function definitions.

**Comment by lmondada** [comment]
> What would be helpful for reviews is to avoid gratuitous symbol name changes in the PR and focus on functional changes or (exclusively or) perfecting the spelling of symbols in the code in separate PRs.

It's tricky. I 100% agree name changes can be confusing and I always try to keep them to a minimum. At the same time there are a lot of function names that don't do what they say. I only changed names where I had to modify code anyway. I'll aim to be even more selective next time.

**Comment by schweitzpgi** [comment]
> > What would be helpful for reviews is to avoid gratuitous symbol name changes in the PR and focus on functional changes or (exclusively or) perfecting the spelling of symbols in the code in separate PRs.
> 
> It's tricky. I 100% agree name changes can be confusing and I always try to keep them to a minimum. At the same time there are a lot of function names that don't do what they say. I only changed names where I had to modify code anyway. I'll aim to be even more selective next time.

Thank you! I appreciate the need for both.

File: python/cudaq/kernel/ast_bridge.py [hunk 5]
```diff
@@ -2530,18 +2509,16 @@ def processQuakeCtor(opName,
...
-            else:
-                decorator = recover_kernel_decorator(name)
-                if decorator is None and name in self.parentVariables:
-                    from .kernel_decorator import isa_kernel_decorator
-                    var = self.parentVariables[name]
-                    if isa_kernel_decorator(var):
-                        decorator = var
 
-            if decorator and not name in self.symbolTable:
+            decorator = recover_value_of_or_none(name, self.defFrame)
```
Comment (atgeller) [inline python/cudaq/kernel/ast_bridge.py:2540]: We'll probably need to make similar changes to https://github.com/NVIDIA/cuda-quantum/blob/63535ff3f58a621d820694b08290873673ec30f2/python/cudaq/kernel/ast_bridge.py#L2639

File: python/cudaq/kernel/kernel_decorator.py (hunk 1 above)
Comment (lmondada) [reply python/cudaq/kernel/kernel_decorator.py:548]: done

File: python/tests/kernel/test_deferred_compilation.py (hunk 2 above)
Comment (lmondada) [reply python/tests/kernel/test_deferred_compilation.py:236]: I've resolved this, too.

File: python/cudaq/kernel/ast_bridge.py (hunk 5 above)
Comment (lmondada) [reply python/cudaq/kernel/ast_bridge.py:2540]: yes, good catch. I will get this in first, and then create a follow up task to sort out any similar bugs in the AST bridge.
